testing_system/code_extractor/code_extractor.py

```diff
@@ -0,0 +1,130 @@
+import ast
+import json
+import re
+from typing import List
+
+from common_util.ast_util import NtbModelInfo, NtbModelInfoAstTraverser
+from common_util.ipynb_grammar import IPYNB_GRAMMAR
+
+
+class NtbCodeExtractor:
+    model_info: List[NtbModelInfo]
+
+    __ntb_specific_patterns = [r'^%']
+
+    def __init__(self, path_to_ntb_file: str):
+        self.path_to_ntb_file = path_to_ntb_file
+        with open(path_to_ntb_file) as ipynb_f:
+            self.ntb_json_content = json.load(ipynb_f)
+        self.model_info = []
+        self.__collect_model_info()
+
+    def get_model_info(self):
+        return self.model_info
+
+    def get_model_names(self):
+        return [mi.model_name for mi in self.model_info]
+
+    def get_preprocessing_code(self) -> str:
+        """
+        By default returns code above constructor of first encountered model
+        """
+        first_model = self.model_info[0]
+        first_model_cell_no, first_model_line_no = first_model.constructor_location.cell_no, \
+                                                   first_model.constructor_location.line_no - 1
+        src_str = '\n'.join(list(self.__generate_code_up_to(first_model_cell_no,
+                                                            first_model_line_no,
+                                                            including_last_line=False)))
+        return src_str
+
+    def get_preprocessing_script(self) -> str:
+        first_model = self.model_info[0]
+
+        res = ''
```

Comment:
Here and on line  78 - why do you creating empty variable beforehand? It looks not very pythonic.
Maybe it will be a good idea to make  in self.__generate_preproc_return_stmt code string argument - it will get as input code and model name and will return already full script.  
As relust you will have something like:


```suggestion
       code = self.__get_preprocessing_code()
       final_script =  self.__generate_preproc_return_stmt(code, first_model.X_train_var_name, 
       first_model.y_train_var_name)
       return final_script
```